src/shaders/splat-shader.ts

```diff
@@ -0,0 +1,170 @@
+
+const vertexShader = /*glsl*/`
+uniform vec3 view_position;
+
+uniform sampler2D splatColor;
+uniform sampler2D splatState;
+uniform highp usampler2D splatTransform;        // per-splat index into transform palette
+uniform sampler2D transformPalette;             // palette of transform matrices
+
+varying mediump vec2 texCoord;
+varying mediump vec4 color;
+flat varying highp uint vertexState;
+#ifdef PICK_PASS
+    flat varying highp uint vertexId;
+#endif
+
+mediump vec4 discardVec = vec4(0.0, 0.0, 2.0, 1.0);
+
+void main(void)
+{
+    // calculate splat uv
+    if (!calcSplatUV()) {
+        gl_Position = discardVec;
+        return;
+    }
+
+    // get center
+    vec3 center = getCenter();
+
+    // get vertex state
+    vertexState = uint(texelFetch(splatState, splatUV, 0).r * 255.0);
+
+    mat4 model = matrix_model;
+
+    // handle per-splat transform
+    uint transformIndex = texelFetch(splatTransform, splatUV, 0).r;
+    if (transformIndex > 0u) {
+        // read transform matrix
+        int u = int(transformIndex % 512u) * 3;
+        int v = int(transformIndex / 512u);
+
+        mat4 t;
+        t[0] = texelFetch(transformPalette, ivec2(u, v), 0);
+        t[1] = texelFetch(transformPalette, ivec2(u + 1, v), 0);
+        t[2] = texelFetch(transformPalette, ivec2(u + 2, v), 0);
+        t[3] = vec4(0.0, 0.0, 0.0, 1.0);
+
+        model = matrix_model * transpose(t);
+    }
+
+    // handle transforms
+    mat4 model_view = matrix_view * model;
+    vec4 splat_cam = model_view * vec4(center, 1.0);
+    vec4 splat_proj = matrix_projection * splat_cam;
+
+    // cull behind camera
+    if (splat_proj.z < -splat_proj.w) {
+        gl_Position = discardVec;
+        return;
+    }
+
+    // get covariance
+    vec3 covA, covB;
+    getCovariance(covA, covB);
+
+    vec4 v1v2 = calcV1V2(splat_cam.xyz, covA, covB, transpose(mat3(model_view)));
+
+    // get color
+    color = texelFetch(splatColor, splatUV, 0);
+
+    // calculate scale based on alpha
+    // float scale = min(1.0, sqrt(-log(1.0 / 255.0 / color.a)) / 2.0);
+
+    // v1v2 *= scale;
```

Comment:
Any need to keep this?

Reply:
Yes, I might reintroduce this optimisation when I have a chance.